wrappers/wrapper.rb

```diff
@@ -742,89 +748,69 @@ def detect_unfeasible_services(vrp)
-            matrix = vrp.matrices.find{ |mat| mat.id == vehicle.matrix_id } if start_index || end_index
+    def service_reachable_by_vehicle_within_timewindows(vrp, activity, vehicle)
+      service_index = activity.point.matrix_index
+      v_start_index = vehicle.start_point&.matrix_index
+      v_end_index = vehicle.end_point&.matrix_index
+      matrix = vrp.matrices.find{ |mat| mat.id == vehicle.matrix_id }
 
-            if vehicle.cost_time_multiplier&.positive? &&
-               !vehicle.cost_late_multiplier&.positive? &&
-               ((vehicle.timewindow&.start && vehicle.timewindow&.end) || vehicle.sequence_timewindows&.size&.positive?)
-              vehicle_available_time = vehicle.timewindow.end - vehicle.timewindow.start if vehicle.timewindow
-              vehicle_available_time = vehicle.sequence_timewindows.collect{ |tw| tw.end - tw.start }.max if vehicle.sequence_timewindows&.size&.positive?
+      vehicle_start = vehicle.timewindow&.start || vehicle.sequence_timewindows.collect(&:start).min || 0
+      vehicle_end = vehicle.timewindow&.end || vehicle.sequence_timewindows.collect(&:end).max || 2**56
+      vehicle_available_time = vehicle_end - vehicle_start
 
-              min_time = (start_index && matrix[:time][start_index][index]).to_f + (end_index && matrix[:time][index][end_index]).to_f + activity.duration
-
-              feasible &&= (vehicle_available_time >= min_time)
-            end
+      earliest_arrival = vehicle_start
+      earliest_arrival += matrix.time[v_start_index][service_index] if v_start_index
+      if !activity.late_multiplier&.positive? && activity.timewindows.any?(&:end)
+        return false if activity.timewindows.all?{ |tw| tw.end && earliest_arrival > tw.end }
+      end
 
-            if feasible &&
-               (start_index || end_index) &&
-               vehicle.cost_distance_multiplier&.positive? &&
-               vehicle.distance
-              min_dist = (start_index && matrix[:distance][start_index][index]).to_f + (end_index && matrix[:distance][index][end_index]).to_f
+      if vehicle_available_time && vehicle.cost_time_multiplier&.positive? && !vehicle.cost_late_multiplier&.positive?
+        earliest_v_end = earliest_arrival + activity.duration
+        earliest_v_end += matrix.time[service_index][v_end_index] if v_end_index
+        return false if (earliest_v_end - vehicle_start) > vehicle_available_time
+      end
 
-              feasible &&= (vehicle.distance >= min_dist)
-            end
+      if !vehicle.cost_late_multiplier&.positive? && activity.timewindows.any?{ |tw| tw.start.positive? } &&
+         (vehicle.timewindow&.end || vehicle.sequence_timewindows.any?(&:end))
+        latest_arrival = vehicle_end - activity.duration
+        latest_arrival -= matrix[:time][service_index][vehicle.end_point.matrix_index] if vehicle.end_point_id
+        return false if activity.timewindows.all?{ |tw| latest_arrival < tw.start }
+      end
```

Comment:
I suspect the first two checks above capture this case, don't they?

Reply:
I do not think it is the case for now but we can edit second check in order to remove this one.

Reply:
I may have mixed some things because we check for vehicle.cost_late_multiplier and then check with service.timewindows. This is not very consistent.

Reply:
In fact timewindow only regards "arrival time" so it is much easier than I thought.

Reply:
oh maybe one of them is for service tw and the other for vehicle tw :/ I suspect they can be cleanly handled ytogether but if not we can keep them seperate. you decide